Assets/Scripts/Mixed/Systems/PlayerPhysicsBasedMovement.cs

```diff
@@ -100,21 +306,49 @@ protected override void OnUpdate()
                 // Make movement vector based on player input
                 float3 movementVelocity = math.mul(horizPlaneView, direction) * speedMultiplier;
 
-                // Preserve current vertical movement
-                movementVelocity.y = pv.Linear.y;
-
                 // Check if is grounded for jumping
-                bool grounded = ISGrounded(trans, settings.maxWalkAngle, settings.groundCheckDistance);
+                float2 groundedCheck = this.AngleBetweenGround(trans.Value, settings.groundCheckDistance, collider, ent.Index, rot.Value);
+                float dist = groundedCheck.x;
+                float angle = groundedCheck.y;
+                bool grounded = dist >= 0 && angle < settings.maxWalkAngle;
+                // Jump if grounded and player hits jump button
                 if (grounded && input.IsJumping)
                 {
-                    movementVelocity.y = settings.jumpForce;
+                    settings.velocity = new float3(0, settings.jumpForce, 0);
                 }
-
-                pv.Angular = float3.zero;
-                pv.Linear = movementVelocity;
+                // Let the player drift to some value very close to the ground
+                else if (!grounded) {
+                    settings.velocity += settings.gravityForce * deltaTime;
+                }
+                // Have hit the ground, stop moving
+                else {
+                    settings.velocity = float3.zero;
+                }
+                float3 finalPos = trans.Value;
+                // Player controlled movement
+                finalPos = ProjectValidMovement(
```

Comment:
nit: rename this first position calculation to something more readable. (It made me look twice on why this was being calculated more than once)

Reply:
Yep, i see what you mean. Breaking this into multiple systems should help.